tools/src/bin/derive-per-file-info.rs

```diff
@@ -0,0 +1,320 @@
+/// ## Output File
+///
+/// For the time being we imitate the bugzilla-components.json representation.
+/// A hierarchical tree representation is retained because the ability to
+/// perform local lookups, walking up the tree as needed, is useful
+/// functionality, as gecko is large enough that global coordination is
+/// impractical, but component level coordination is feasible.
+///
+/// - "bugzilla-components": Directly from bugzilla-components.json, a
+///   dictionary mapping from stringified numeric values to list tuples of the
+///   form [product, component].
+/// - "root": More explicit version of bugzilla-components.json's rep.  Each
+///   node in the tree structure is an object with one of the following forms,
+///   with the root having type "dir":
+///   - { type: "dir", contents } where:
+///     - "contents": An object directionary whose keys are filenames and value
+///        nodes.
+///   - { type: "file", component, testInfo } where:
+///     - "component": Value is the numeric bugzilla component to be looked up.
+///     - "testInfo": The value nodes from `test-info-alltests.json` verbatim.
+///       In the future this may gain more data.
+///     - "wptInfo": A digested version of the per-test info.  The presence of
+///       this field means that a per-test metadata file exists which is itself
+///       an indication of there being some kind of notable manipulation going
+///       on even if we don't understand it specifically.
+///       - "disabled": Directly propagated from the test-level "disabled".
+fn main() {
+    env_logger::init();
+
+    let args: Vec<_> = env::args().collect();
+
+    let cfg = config::load(&args[1], true);
+    println!("Config file read");
+
+    let tree_name = &args[2];
+    let tree_config = cfg.trees.get(tree_name).unwrap();
+
+    // ## Build empty derived info structure
+    let mut all_info: json::Object = BTreeMap::new();
+    {
+        let contents = json::Object::new();
+        let mut root = json::Object::new();
+        root.insert("type".to_string(), "dir".to_string().to_json());
+        root.insert("contents".to_string(), Json::Object(contents));
+        all_info.insert("root".to_string(), Json::Object(root));
+    }
+
+    // ## Load bugzilla data and merge it in to the derived info structure
+    let bugzilla_fname = format!("{}/bugzilla-components.json", tree_config.paths.index_path);
+    let bugzilla_data = read_json_from_file(&bugzilla_fname);
+    if let Some(mut data) = bugzilla_data {
+        println!("Bugzilla components read");
+
+        all_info.insert("bugzilla-components".to_string(),
+                        data.remove("components").unwrap());
+
+        if let Some(bz_root) = data.get("paths") {
+            traverse_and_store_bugzilla_map(
+                bz_root.as_object().unwrap(),
+                all_info.get_mut("root").unwrap().as_object_mut().unwrap());
+        }
+    } else {
+        println!("No bugzilla-components.json file found");
```

Comment:
Should we make this a `warn!` instead so we get notified of problems? Ditto other missing files. And I guess the rest of the `println!` calls should really be `info!`s but that's a pre-existing problem so no need to fix that here.